src/js/searches/programs.ts

```diff
@@ -14,7 +17,47 @@ export function filenameCorrelation(md5: string) {
   return `md5=${md5} | count() by filename, mime_type | sort -r | head 5`
 }
 
+export function uidFilter(uid: string | zng.Primitive) {
+  return zql`uid=${uid} or ${uid} in conn_uids or ${uid} in uids or referenced_file.uid=${uid}`
+}
+
+export function cidFilter(cid: string) {
+  return `community_id="${cid}"`
+}
+
 export const UID_CORRELATION_LIMIT = 100
+export function correlationLimit() {
+  return `head ${UID_CORRELATION_LIMIT}`
+}
+
+type RelatedIds = {
+  uid?: string
+  cid?: string
+}
+
+export function correlationIds({uid, cid}: RelatedIds) {
+  const filters = []
+  if (uid) filters.push(uidFilter(uid))
+  if (cid) filters.push(cidFilter(cid))
+  return [filters.join(" or "), correlationLimit()].join(" | ")
+}
+
 export function uidCorrelation(uid: string) {
-  return `uid=${uid} or ${uid} in conn_uids or ${uid} in uids or referenced_file.uid=${uid} | head ${UID_CORRELATION_LIMIT}`
+  return `${uidFilter(uid)} | ${correlationLimit()}`
+}
+
+export function cidCorrelation(cid: string) {
+  return `${cidFilter(cid)} | ${correlationLimit()}`
+}
+
+export function connCorrelation(conn: zng.Record) {
+  const uid = conn.get("uid") as zng.Primitive
+  const cid = conn.get("community_id") as zng.Primitive
+  const ts = (conn.get("ts") as zng.Primitive).toDate()
+  const dur = (conn.get("duration") as zng.Primitive).toFloat() + 90 // Add a 1.5 minute buffer for events that get logged late
+  const endTs = new Date(new Date(ts).getTime() + dur * 1000)
+
+  const cidFilter = zql`community_id = ${cid} and ts >= ${ts} and ts < ${endTs}`
```

Comment:
Ok, so I'm pretty stoked about this. I'm using a template literal function for the first time. It allows us to interpolate complex objects into the zql string and smartly translate them into valid zql. 

For example, we can pass in a JS Date object and it will convert it to time.

```js
const ts = new Date(123)
const query = zql`ts >=${ts}`

console.log(query)
// "ts >= 00000123.0"
```

We could even pass whole zng.DataTypes and it could convert them to their equivalent zql strings.

```js
const uid = conn.get("uid") // => zng.Primitive class
zql`uid={uid}` => 'uid="abc123k48%@4kdfsf"'
```

Reply:
🔥